Newtype derivable codec class

clintonmead opened the pull request:
I'm just making this PR to keep you up to date with a fork of `autodocodec` I'm maintaining.

I know I submitted some previous PRs without tests (https://github.com/NorfairKing/autodocodec/pull/44) and never got around to adding those tests, so sorry for being slack about this.

Unfortunately the situation isn't much better with this PR, except now I've significantly changed the definition of the `Codec` class, to make it's parameters of representation role, which means `HasCodec` can now be newtyped derived (and also `Codec` can be `coerce`d). 

This brings `HasCodec` in line with classes like `ToJSON`, `FromJSON` and `ToSchema`, which can be newtype derived.

The downside to this is lots of new `Coercible` constraints in the definition of `Codec` and a bunch of new calls to `coerce`. 

I understand this is a controversial change which you may not want to accept, it does complicate the code for main benefit of being able to use `deriving newtype HasCodec` instead of:

```
instance HasCodec T where
  codec = dimapCodec T fromT codec
```

But I've got so many newtypes in my application this convenience was worth it to me. 

But I thought it best to at least keep you up to date with the changes I'm making, you may want to merge at least some of them.

I will try to get around to adding tests for the new instances also (although with this change, the new instances are actually just newtype derived, so perhaps the tests are less needed). But I don't think I'll get around to that for at least a month or so.

NorfairKing commented on autodocodec/src/Autodocodec/Codec.hs:
```diff
@@ -73,53 +75,61 @@ import qualified Data.Aeson.KeyMap as KM
+
+type role Codec _ representational representational
 data Codec context input output where
   -- | Encode '()' to the @null@ value, and decode @null@ as '()'.
-  NullCodec ::
-    ValueCodec () ()
+  NullCodec :: 
+    (Coercible a (), Coercible b ()) => 
+    ValueCodec a b
```
I do not understand why this works at all. Could you elaborate?

clintonmead replied:
The issue is that `Codec`, as it currently exists, has `nominal` roles for the parameters `input` and `output`. If we look at [Coercible](https://hackage.haskell.org/package/base-4.20.0.1/docs/Data-Coerce.html#t:Coercible) one can only coerce `T a` to `T b` if `Coercible a b` is true AND the role of the first parameter is representational. And as a consequence, we need the roles of our parameters to be representational if we want to do newtype deriving (as what newtype deriving does is just coercions (which are quite nicely, completely free at runtime).

How do we make the roles of these parameters representational? Well looking at the documentation on [roles](https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/roles.html#role-inference) we have this perhaps vague statement:

> all GADT-like parameters are nominal

The docs don't specify what "GADT-like" parameters are, but basically they are, as far as I understand, any GADT  constructor where the RHS isn't completely free in the parameter.

For example:

```
T :: ... -> T a b
```

Doesn't have a GADT style parameter but:

```
T :: ... -> T Int Int
T :: ... -> T a a
T :: C a => ... -> T a b
```

are GADT style.

Any constructor which has a GADT style parameter forces the parameter to be nominal for the whole datatype.

Of course, even if you have:

```
T :: U a b -> T a b
```

And `U`s parameters are nominal, then that will be inherited (this isn't an issue for `Codec`).

So we need to make all the constructors which have values of `input` and `output` which aren't free, to be free.

So for example, with `NullCodec`, instead of `ValueCodec () ()`, we do `(Coercible a (), Coercible b ()) => ValueCodec a b`.

This is really just a generalisation of `NullCodec`, because we could actually write:

```
nullCodec :: (Coercible a (), Coercible b ()) => ValueCodec a b
nullCodec = dimapCodec coerce coerce NullCodec
```

anyway. So doing this refactoring shouldn't break any invariants.

Note that the statement in the [roles docs](https://ghc.gitlab.haskell.org/ghc/doc/users_guide/exts/roles.html#role-inference) that says:

> the default role for classes is nominal

The only exception to this is `Coercible`, which role is representational. 

So by doing this to all the constructors that don't have their parameters free in both parameters, we then make `Codec`'s parameters have representational roles.

Then we can do newtype deriving.

We get some added complexity internally but we give clients more power by giving them the ability to coerce codecs and newtype derive instances.

Including the newtyped instance below.

These would not be possible with the current definition of `Codec`.

NorfairKing replied:
@clintonmead I'm not convinced yet that this is worth it, but I'm very curious about how this all fits together.
Why can't GHC do this for us?

clintonmead replied:
@NorfairKing "Why can't GHC do this for us?"

Because there are good reasons why one would want nominal parameters.

Basically what this transformation does is make the parameters representative, which means for a given data type `T`, makes:

```
Coercible a b => Coercible (T a) (T b)
```

There are some cases where we do not want to do this. Consider the type [Down](https://hackage.haskell.org/package/base-4.20.0.1/docs/Data-Ord.html#t:Down) and [SortedList](https://hackage.haskell.org/package/sorted-list-0.2.2.0/docs/Data-SortedList.html). 

`Down` exposes it's constructor. In that `coerce :: a -> Down a` is valid.

Should `coerce :: SortedList a -> SortedList (Down a)` be valid?

Certainly not. `SortedList` has an invariant that the elements are sorted based on the `Ord` instance of the type. If you just coerce ` SortedList a -> SortedList (Down a)` the elements are now in the wrong order, violating the invariant of the class.  (Indeed, looking at the source of `SortedList`, I think it's actually buggy and should have a role annotation to prevent this bad coercion). 

There's other circumstance where this transformation just complicate things. For example if you're just making an AST that represents your language, in that you have "Numbers", "Symbols", "Ops" etc. Your list types of expressions are finite and enumerable. You want `Expr Number` or `Expr Symbol` etc, there's no need for the added complexity of `Coercible a Num => Expr a`. 

But I think in the case of `Codec`, there's a few points in favour of this transformation:

1. There's no internal invariants this transformation is violating. This is evidenced by the fact that you could just do `dimap coerce coerce` and achieve the same thing already.
2. Coercing, and in particular, newtype deriving codecs is really useful. Indeed `ToJSON` and `FromJSON` allow for this.

NorfairKing commented on autodocodec/src/Autodocodec/Class.hs:
```diff
@@ -117,10 +128,16 @@ instance HasCodec Natural where
 
-instance (HasCodec a) => HasCodec (Identity a) where
-  codec = dimapCodec runIdentity Identity codec
+deriving newtype instance HasCodec a => HasCodec (Identity a)
+deriving newtype instance HasCodec (f a) => HasCodec (Alt f a)
+deriving newtype instance HasCodec a => HasCodec (Dual a)
+deriving newtype instance HasCodec a => HasCodec (Semigroup.First a)
+deriving newtype instance HasCodec a => HasCodec (Semigroup.Last a)
+deriving newtype instance HasCodec a => HasCodec (Monoid.First a)
+deriving newtype instance HasCodec a => HasCodec (Monoid.Last a)
+deriving newtype instance HasCodec a => HasCodec (Const a b)
```
I wouldn't mind some of these, but with tests.

clintonmead commented:
Another commit. Because the `HasCodec` instances for things like `UTCTime` were just `bimapCodec` wrapped around `ValueCodec`, the underlying type was seen as a JSON Value, not a string, which was mucking up the OpenAPI generation.

I've changed these functions to wrap `StringCodec` with the [`iso8601ParseM`](https://hackage.haskell.org/package/time-1.14/docs/Data-Time-Format-ISO8601.html#v:iso8601ParseM) and [`iso8601Show`](https://hackage.haskell.org/package/time-1.14/docs/Data-Time-Format-ISO8601.html#v:iso8601Show) functions from the [`time` library](https://hackage.haskell.org/package/time-1.14/docs/Data-Time-Format-ISO8601.html).

This _should_ fix the OpenAPI generation for these instances although I'm yet to confirm this.

NorfairKing commented:
@clintonmead Try running `nix flake check` and fix things until that succeeds.

NorfairKing commented:
@clintonmead I would really like to accept your contributions but I need your help to get them in shape.
In particular:
* `nix flake check` needs to be passing for each commit
* We need tests for each instance/feature. Most of that work has already been done. It just involves adding lines in the right places and running the test suite with `--golden-reset`.
* Each contribution needs to be separated out into a nice commit.
* Each change needs to be reflected in the changelog.

Could you do these?
Otherwise these PRs might stay here indefinitely.

clintonmead commented:
@NorfairKing I've done everything except for "We need tests for each instance/feature... It just involves adding lines in the right places..."

Where are the "right places"? Are these just the doc strings inline with the file?

NorfairKing commented:
> @NorfairKing I've done everything except for "We need tests for each instance/feature... It just involves adding lines in the right places..."
> 
> Where are the "right places"? Are these just the doc strings inline with the file?

The api usage tests need to be updated in two ways:

1. In `src` we need to add something that wouldn't have compiled before and now does
2. It `test` we need to add tests for all the instances that we add in each of the testing modules.

clintonmead commented:
@NorfairKing by `src` and `test` do you mean:

1. `autodocodec-api-usage/src/Autodocodec/Usage.hs`
2. `autodocodec-api-usage/test/Autodocodec/AesonSpec.hs`

as most relevant to my changes and additions?

NorfairKing commented:
@clintonmead 

> @NorfairKing by `src` and `test` do you mean:
> 
>     1. `autodocodec-api-usage/src/Autodocodec/Usage.hs`

This is right.

>     2. `autodocodec-api-usage/test/Autodocodec/AesonSpec.hs`

No, all of these:

```
autodocodec-api-usage/test/Autodocodec/YamlSpec.hs
autodocodec-api-usage/test/Autodocodec/AesonSpec.hs
autodocodec-api-usage/test/Autodocodec/SwaggerSpec.hs
autodocodec-api-usage/test/Autodocodec/Aeson/SchemaSpec.hs
autodocodec-api-usage/test/Autodocodec/ShowSpec.hs
autodocodec-api-usage/test/Autodocodec/MultipartSpec.hs
autodocodec-api-usage/test/Autodocodec/OpenAPISpec.hs
autodocodec-api-usage/test/Autodocodec/Yaml/SchemaSpec.hs
```

It'll be one line per instance in each of these.

clintonmead commented:
@NorfairKing I've made some progress on adding tests (I've repushed the PR). Could you sanity check my additions and tell me if they're not what you're looking for?

Also I tried running `stack test --test-arguments="--golden-reset"` to generate fresh test data but it didn't seem to generate anything. Am I using the wrong command or is there something still missing in my PR to generate fresh target test data?

NorfairKing commented:
@clintonmead I'll have a look.
You'll need `--golden-start` to generate new files and `--golden-reset` to regenerate them :)

NorfairKing commented on autodocodec-api-usage/test/Autodocodec/Orphans.hs:
```diff
@@ -0,0 +1,50 @@
+import qualified Data.Semigroup as Semigroup
+import Test.Syd.Validity
+
+deriving newtype instance (Validity a) => Validity (Identity a)
```
These need to be upstreamed to `validity` and `genvalidity` at https://github.com/NorfairKing/validity first.

clintonmead replied:
@NorfairKing I've created a PR for this here: https://github.com/NorfairKing/validity/pull/119

NorfairKing replied:
Merged and released.

NorfairKing commented on autodocodec/src/Autodocodec/Class.hs:
```diff
@@ -144,31 +165,29 @@ instance (Ord a, HasCodec a) => HasCodec (Set a) where
--- TODO make these instances better once aeson exposes its @Data.Aeson.Parser.Time@ or @Data.Attoparsec.Time@ modules.
 instance HasCodec Day where
-  codec = codecViaAeson "Day"
+  codec = codecViaISO8601 "Day"
```
I wonder if the roundtrip tests with aeson start failing after this change.

clintonmead replied:
I haven't noticed this causing issues in the test runs.

clintonmead replied:
It seems to work currently, but I see the point that we want to keep consistency with aeson. 

I've instead changed the approach as a result. I think I really should put a PR to aeson, but in the meantime I think the solution I've put now is better, even if it's a bit scary.

clintonmead commented:
@NorfairKing I _think_ I've addressed all the tests now, shout out if I've missed something or want any further changes. Thanks for merging the validity change so quickly. 

I've actually removed a both the `()` and `Alt` codecs as compared to the original PR, because I wasn't using them anyway now, and `Alt` doesn't have an `aeson` codec, and `aeson` strangely serialises `()` as `Array []`. I then thought there isn't an obvious canonical choice of codec for `()` so I just removed it.

NorfairKing commented on autodocodec-api-usage/test_resources/json-schema/utc-time.json:
```diff
@@ -1,3 +1,4 @@
 {
-    "$comment": "LocalTime"
+    "$comment": "UTCTime",
+    "type": "string"
```
That's indeed better!

NorfairKing commented on autodocodec-api-usage/test_resources/nix/day-type.nix:
```diff
@@ -1 +1 @@
-types.unspecified
+types.str
```
I just realised you put this on top of the `autodocodec-nix` branch.
I guess that means I really have to finish that library now :P
(This is good!)

NorfairKing commented on autodocodec/src/Autodocodec/Codec.hs:
```diff
@@ -1880,3 +1911,36 @@ codecViaAeson ::
+instance MonadFail Result where
+  fail = Result . Left
+
+-- TODO 'aeson' has it's own custom datetime serialising code in the module @Data.Aeson.Encoding.Builder@:
+-- The core function here is `Data.Aeson.Encoding.Builder.timeOfDay64`.
+-- However, this module is private.
+-- There is @Data.Aeson.Encoding.Internal@, which interestingly isn't private, but it's only exposed functions
+-- wrap the return bytestring in a quotes. Only for the quotes to be removed in `Data.Aeson.Types.ToJSON.stringEncoding`
+-- This all seems a bit silly.
+-- I think `aeson` should just expose @Data.Aeson.Encoding.Builder@ and it's datetime instances should just take those builders,
+-- convert them to Text and be done with it.
+-- I plan to submit a PR to 'aeson' to do this.
+-- In the meantime, I think the best way to ensure we are exactly behaving as 'aeson' is just to _assume_ aeson is returning a string
+-- This is a correct assumption for any of the datetime types, but using this function generally is unsafe.
```
Hehe, looks like you ran into why I used `codecViaAeson` :D
What's left `TODO` here?

clintonmead replied:
TODO is to submit a PR to aeson and then adjust this library to use the patched aeson code

NorfairKing replied:
@clintonmead do you want to do that before I merge?

clintonmead replied:
@NorfairKing there was previously a todo regarding awaiting an upstream change to the aeson library, I've just elaborated some more. You can remove this more detailed comment and go back to the briefer comment if you would prefer, but I was just following the libraries previous practice of writing todos regarding changes required in dependent libraries.

NorfairKing replied:
Ok cool! Then this PR is ready, right?

NorfairKing commented:
I'm about to merge this. Thank you @clintonmead

NorfairKing commented:
Merged manually.
